lib/puppet/provider/ios_ntp_config/cisco_ios.rb

```diff
@@ -0,0 +1,64 @@
+require_relative '../../../puppet_x/puppetlabs/cisco_ios/check'
+require_relative '../../util/network_device/cisco_ios/device'
+require_relative '../../../puppet_x/puppetlabs/cisco_ios/utility'
+
+# NTP Config Puppet Provider for Cisco IOS devices
+class Puppet::Provider::IosNtpConfig::CiscoIos
+  def self.commands_hash
+    local_commands_hash = PuppetX::CiscoIOS::Utility.load_yaml(File.expand_path(__dir__) + '/command.yaml')
+    ntp_config_commands_hash = PuppetX::CiscoIOS::Utility.load_yaml(File.expand_path(__dir__) + '/../ntp_config/command.yaml')
+    @commands_hash = local_commands_hash.merge(ntp_config_commands_hash) { |_key, oldval, newval| (oldval.to_a + newval.to_a).to_h }
+    @commands_hash
+  end
+
+  def self.instances_from_cli(output)
+    new_instance_fields = []
+    new_instance = PuppetX::CiscoIOS::Utility.parse_resource(output, commands_hash)
+    new_instance[:name] = 'default'
+    new_instance[:update_calendar] = (new_instance[:update_calendar]) ? true : false
+    new_instance.delete_if { |_k, v| v.nil? }
+    new_instance_fields << new_instance
+    new_instance_fields
+  end
+
+  def self.commands_from_is_should(_is, should)
+    array_of_commands = []
+    should.delete(:name)
+    should[:update_calendar] = 'unset' unless should[:update_calendar]
+    array_of_commands += PuppetX::CiscoIOS::Utility.build_commmands_from_attribute_set_values(should, commands_hash)
+    array_of_commands
+  end
+
+  def commands_hash
+    Puppet::Provider::IosNtpConfig::CiscoIos.commands_hash
+  end
+
+  def get(context, _names = nil)
+    output = context.transport.run_command_enable_mode(PuppetX::CiscoIOS::Utility.get_values(commands_hash))
+    return [] if output.nil?
+    return_value = Puppet::Provider::IosNtpConfig::CiscoIos.instances_from_cli(output)
+    PuppetX::CiscoIOS::Utility.enforce_simple_types(context, return_value)
+  end
+
+  def set(context, changes)
+    changes.each do |name, change|
+      is = change.key?(:is) ? change[:is] : (get(context) || []).find { |key| key[:name] == name }
+      should = change[:should]
+
+      context.updating(name) do
+        update(context, name, is, should)
+      end
+    end
+  end
+
+  def update(context, _name, is, should)
+    array_of_commands_to_run = Puppet::Provider::IosNtpConfig::CiscoIos.commands_from_is_should(is, should)
+    array_of_commands_to_run.each do |command|
+      context.transport.run_command_conf_t_mode(command)
+    end
+  end
+
+  def canonicalize(_context, resources)
+    resources
```

Comment:
There should be a sort in here for the trusted keys, in order to maintain idempotency.

Reply:
You still need a sort in here fr trusted_keys to maintain idempotency. The entire array is compared for the check, not the individual items. i.e. If you change from:
```
ios_ntp_config { 'default':
  authenticate => true,
  source_interface => 'Vlan42',
  trusted_key => [12,24,48,96],
  update_calendar => true,
}
```
to
```ios_ntp_config { 'default':
  authenticate => true,
  source_interface => 'Vlan42',
  trusted_key => [12,48,24,96],
  update_calendar => true,
}
```
you will get:
```
Notice: /Stage[main]/Main/Ios_ntp_config[default]/trusted_key: trusted_key changed [12, 24, 48, 96] to [12, 48, 24, 96]
```
This sort of thing is what the canonicalize is added for. Try something like:
```
    resources.each do |resource|
      resource[:trusted_keys] = resource[:trusted_keys].sort if resource[:trusted_keys]
    end

```